### DISCUSSION TITLE
modbus LAN addresses

### DISCUSSION THREAD

**Discussion by TonyM1958** [StealthChesnut/HA-FoxESS-Modbus#116] [General]
Did some looking at the other modbus values presented by the inverter to see if I could find any of the BMS values present on RS-485. Early days, but it appears only addresses 31000 to 31039 are active as requesting any more than 40 registers gets rejected by the server. Started putting some additional values in the attached spread sheet. Maybe there is another block of addresses that I haven't stumbled across yet.

31010 to 31013: my guess is that these might be EPSVolt, EPSCurrent, EPSPower and EPSFreq? My inverter is in internal bypass mode and it looks like it measures my EPS voltage and frequency but not the current or power. Obviously, these won't tell you anything if you are on external bypass. So these registers might only useful when the grid is down and the inverter is generating EPS power, assuming you have something plugged in that can still read them!

31028:  suggestion that this might be work mode, but I switched between self-use to feed-in first and it didn't change, so I don't think it is that? Could just be other [static?] data from 31025 to 31039.

[H1 LAN modbus map.xlsx](https://github.com/StealthChesnut/HA-FoxESS-Modbus/files/11091248/H1.LAN.modbus.map.xlsx)

**Comment by TonyM1958** [comment]
@StealthChesnut @williamjeccles I hooked up Modbus Tool to RS485 and did a scan of my inverter. Looks like there are up to 50 registers as per the attached screen shot, so I'm looking at trying to fill in some of the gaps and align more of the data between RS485 and LAN

![modbus rs485 scan](https://user-images.githubusercontent.com/63789168/228823417-13a0db35-2f9a-456e-aba3-05b74613a051.png)

I was looking for InvBatCurrent that is available for LAN but not for RS485. I've added some guesses at other registers in the attached spreadsheet. If you already know what any of the addresses are, please can you let me know so I don't duplicate or go on a wild goose chase?

[H1 modbus map.xlsx](https://github.com/StealthChesnut/HA-FoxESS-Modbus/files/11110951/H1.modbus.map.xlsx)

**Comment by StealthChesnut** [comment]
Hi Tony,

Great stuff - thanks for taking a look at this. I assume you've already seen the reference, but in case you haven't - https://github.com/StealthChesnut/HA-FoxESS-Modbus/wiki/Data-Register-Reference---H1-AC1

@nathanmarlor has also made some progress in his integrations here: https://github.com/nathanmarlor/foxess_modbus/blob/main/custom_components/foxess_modbus/entities/xx1_aux_entity_descriptions.py
Worth checking out.

**Reply by TonyM1958** [reply]
Thanks - yes, I started by plagiarising your work from the wiki :-)
Just after I posted my previous comment, it occured to me to have a look at what @nathanmarlor had done as well ;-)
If Fox won't publish their modbus map, perhaps we can come up with a reasonably complete one between us.

**Comment by TonyM1958** [comment]
Had a skim through entity code by @nathanmarlor and updated the spread sheet as attached.
[H1 modbus map.xlsx](https://github.com/StealthChesnut/HA-FoxESS-Modbus/files/11113803/H1.modbus.map.xlsx)

For completeness - I'm not looking at H3 as I don't have an example to poke around with.

I'd be interested in any views on trying to get better alignment between entity names so, for example, users can switch between LAN and RS485 connections easily and retain their history. It would be a breaking change, so would be focussed on LAN where the current number of users might be quite low due to the problems caused by firmware 1.56. There's also some oddities like PV1-Voltage (@StealthChesnut) and PV1_voltage (@nathanmarlor) , BatVolt / InvBatVolt, CT2 / CT2_Meter...

Can you setup aliases in HA to retain backward compatibility but achieve better alignment?

Or is this just a bridge too far :-)

**Reply by nathanmarlor** [reply]
The entity IDs match so history can be retained when switching between the two. The descriptions may be different but this doesn't impact compatibility.

**Reply by TonyM1958** [reply]
thanks Nathan. I haven't dug much into the internals of HA or your code (yet). I have Stealthchesnut installed for my dashboard but am, hopefully like others, thinking about changing at some point. Ideally, this would be seamless but, as far as I know, my current setup does not have entity IDs for modbus (at least that's what HA tells me if I try to change one). Are your entity IDs static and would it be worth back porting them into the Stealthchesnut setup to align names and history and make moving easier in the future?

I'm happy to work on documenting this and adding to the spread sheet if that helps.

**Reply by nathanmarlor** [reply]
@TonyM1958 that's exactly what I did when I created the integration :) because Chris' integration doesn't specify unique IDs HA uses the description (i.e. PV1 Power becomes sensor.pv1_power). I've therefore matched these unique IDs but in some cases provided a different description given what we have learnt.

Moving between the two was designed to be seamless without loss of history etc. @FozzieUK has provided a description here - https://github.com/nathanmarlor/foxess_modbus/issues/87 which I'll merge into the readme asap.

**Reply by TonyM1958** [reply]
thanks, will have a look. It appears there has been some divergence over time, so possibly things just need tightening up? I didn't need to do a lot to flip beween @StealthChesnut LAN and USB connections but there definitely were a few disconnects. Would it help if I go back and document these?

Stiil learning, but I spent a while recently doing SQL, python and AI stuff in Jupyter Lab. I've added a few more templates, largely for presentation purposes if anyone wants to steal (I know, more comments required...):

```
template:
  - sensor:
      - name: "inverter_efficiency"
        unit_of_measurement: "%"
        state: >
          {% set power_out = states('sensor.inverter_power_out') | float(default=0) %}
          {% set power_in = states('sensor.inverter_power_in') | float(default=0) %}
          {% set efficiency = power_out / power_in * 100 if power_in > 0 else null %}
          {% set efficiency = 100 if efficiency > 100 else efficiency %}
          {{ efficiency | round(1) }}
      - name: "inverter_efficiency_daily"
        unit_of_measurement: "%"
        state: >
          {% set power_out = states('sensor.inverter_out_daily') | float(default=0) %}
          {% set power_in = states('sensor.inverter_in_daily') | float(default=0) %}
          {% set efficiency = power_out / power_in * 100 if power_in > 0 else null %}
          {% set efficiency = 100 if efficiency > 100 else efficiency %}
          {{ efficiency | round(1) }}
      - name: "battery_cell_imbalance"
        unit_of_measurement: "%"
        state: >
          {% set cell_high = states('sensor.bms_cell_mv_high') | float(default=0) %}
          {% set cell_low = states('sensor.bms_cell_mv_low') | float(default=0) %}
          {% set imbalance = (cell_high-cell_low) / cell_low * 100 if cell_low > 0 else null %}
          {{ imbalance | round(2) }}
      - name: "battery_duration"
        icon: "mdi:timer-play"
        unit_of_measurement: "h"
        state: >
          {% set min_soc = 4 * 2.56 * 0.12 %}
          {% set energy = states('sensor.bms_kwh_remaining') | float(default=0) %}
          {% set power = states('sensor.load_power') | float(default=0) %}
          {% set power = 0.1 if power < 0.1 else power %}
          {% set hours = ((energy - min_soc) / power) | int  %}
          {% set hours = 24 if hours > 24 else hours  %}
          {% set hours = 0 if hours < 0 else hours  %}
          {{ hours }}
      - name: "battery_backfill"
        device_class: "energy"
        unit_of_measurement: "kWh"
        state: >
          {{( states('sensor.pv_energy_daily')|float(default=0)
            + states('sensor.grid_daily')|float(default=0)
            - states('sensor.load_daily')|float(default=0)
            - states('sensor.feedin_daily')|float(default=0)
            - states('sensor.losses_daily')|float(default=0)
            ) | round(3) }}
      - name: "inverter_power_in"
        device_class: "power"
        unit_of_measurement: "kW"
        state: >
          {{( states('sensor.pv_power_now') | float(default=0)
            + states('sensor.grid_consumption') | float(default=0)
            + states('sensor.battery_discharge') | float(default=0)
            ) | round(3) }}
      - name: "inverter_power_out"
        device_class: "power"
        unit_of_measurement: "kW"
        state: >
          {{( states('sensor.feed_in_power') | float(default=0)
            + states('sensor.load_power') | float(default=0)
            + states('sensor.battery_charge') | float(default=0)
            ) | round(3) }}
      - name: "net_import_daily"
        device_class: "power"
        unit_of_measurement: "kWh"
        state: >
          {{( states('sensor.grid_daily') | float(default=0)
            - states('sensor.feedin_daily') | float(default=0)
            ) | round(3) }}
      - name: "grid_dependency_daily"
        device_class: "power"
        unit_of_measurement: "kWh"
        state: >
          {{( states('sensor.load_daily') | float(default=0)
            - states('sensor.pv_energy_daily') | float(default=0)
            ) | round(3) }}
```
